From 0a6e6924e7250c88f13faeb245fbe666c70b65d7 Mon Sep 17 00:00:00 2001
From: Cheng-Cheng Lo <locc@google.com>
Date: Thu, 21 Aug 2025 16:50:01 +0800
Subject: [PATCH 17/55] Disallow factory reset while in DSU mode

Bug: 430568718
Flag: EXEMPT CVE_FIX
Test: presubmit
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:72f795b3bcb0c98971f5f15689c9399c8160bfcb)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:0846e0773b3f2fc54ec2e9674f55115f84d64e1e
Merged-In: I36afa074f16bb7ebf8b617de73dab3d4df5fb5dc
Change-Id: I36afa074f16bb7ebf8b617de73dab3d4df5fb5dc
---
 core/java/android/os/RecoverySystem.java | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/core/java/android/os/RecoverySystem.java b/core/java/android/os/RecoverySystem.java
index 398140daf..709cb3fed 100644
--- a/core/java/android/os/RecoverySystem.java
+++ b/core/java/android/os/RecoverySystem.java
@@ -37,6 +37,7 @@ import android.content.IntentFilter;
 import android.content.IntentSender;
 import android.content.pm.PackageManager;
 import android.hardware.display.DisplayManager;
+import android.os.image.DynamicSystemManager;
 import android.provider.Settings;
 import android.telephony.SubscriptionInfo;
 import android.telephony.SubscriptionManager;
@@ -934,9 +935,15 @@ public class RecoverySystem {
     public static void rebootWipeUserData(Context context, boolean shutdown, String reason,
             boolean force, boolean wipeEuicc, boolean keepMemtagMode) throws IOException {
         UserManager um = (UserManager) context.getSystemService(Context.USER_SERVICE);
+        DynamicSystemManager dsuManager = (DynamicSystemManager)
+                    context.getSystemService(Context.DYNAMIC_SYSTEM_SERVICE);
         if (!force && um.hasUserRestriction(UserManager.DISALLOW_FACTORY_RESET)) {
             throw new SecurityException("Wiping data is not allowed for this user.");
         }
+
+        if (dsuManager.isInUse()) {
+            throw new SecurityException("Wiping data is not allowed while in DSU mode.");
+        }
         final ConditionVariable condition = new ConditionVariable();
 
         Intent intent = new Intent("android.intent.action.MASTER_CLEAR_NOTIFICATION");
-- 
2.52.0

